src/dmd/escape.d

```diff
@@ -154,71 +154,52 @@ bool checkMutableArguments(Scope* sc, FuncDeclaration fd, TypeFunction tf,
-                            if (!gag)
-                            {
-                                const(char)* msg = eb.isMutable && eb2.isMutable
-                                    ? "more than one mutable reference of `%s` in arguments to `%s()`"
-                                    : "mutable and const references of `%s` in arguments to `%s()`";
-                                error((*arguments)[i].loc, msg,
-                                    v.toChars(),
-                                    fd ? fd.toPrettyChars() : "indirectly");
-                            }
-                            errors = true;
-                        }
-                    }
-                }
-            }
+            const(char)* msg = eb.isMutable && eb2.isMutable
+            ? "more than one mutable reference of `%s` in arguments to `%s()`"
+            : "mutable and const references of `%s` in arguments to `%s()`";
+            error((*arguments)[i].loc, msg,
+                  v.toChars(),
+                  fd ? fd.toPrettyChars() : "indirectly");
         }
+        errors = true;
+    }
 
-        foreach (VarDeclaration v; eb.er.byref)
+    void escape(string by)(size_t i, ref EscapeBy eb, bool skipPointers)
+    {
+        foreach (VarDeclaration v; mixin("eb.er."~by))
         {
-            if (log) printf("byref `%s`\n", v.toChars());
+            if (log) printf( (by ~ "`%s`\n\0").ptr, v.toChars());
+            if (skipPointers && !v.type.hasPointers())
+                continue;
             foreach (ref eb2; escapeBy[i + 1 .. $])
             {
-                foreach (VarDeclaration v2; eb2.er.byref)
+                foreach (VarDeclaration v2; mixin("eb2.er."~by))
                 {
-                    if (log) printf("v2: `%s`\n", v2.toChars());
-                    if (v2 != v)
-                        continue;
-                    //printf("v %d v2 %d\n", eb.isMutable, eb2.isMutable);
-                    if (eb.isMutable || eb2.isMutable)
-                    {
-                        if (global.params.vsafe && sc.func.setUnsafe())
-                        {
-                            if (!gag)
-                            {
-                                const(char)* msg = eb.isMutable && eb2.isMutable
-                                    ? "more than one mutable reference to `%s` in arguments to `%s()`"
-                                    : "mutable and const references to `%s` in arguments to `%s()`";
-                                error((*arguments)[i].loc, msg,
-                                    v.toChars(),
-                                    fd ? fd.toPrettyChars() : "indirectly");
-                            }
-                            errors = true;
-                        }
-                    }
+                    checkOnePair(i, eb, eb2, v, v2);
                 }
             }
         }
     }
+    foreach (const i, ref eb; escapeBy[0 .. $ - 1])
+    {
+        escape!"byvalue"(i, eb, true);
```

Comment:
Why the mixin ? You could just pass a `ref` to `byref` / `byvalue` no ?

Reply:
while eb is fine, eb2 in `escape` is not, it needs to know which field to access. I suppose I could do it with `.offsetof` or `.tupleof` but I'd rather use a mixin than those. If you have better idea please do tell.

Reply:
```
foreach (VarDeclaration v2; by == byref ? eb2.er.byref : eb2.er.byvalue)
```